File: appkit-creating-and-customizing-the-touch-bar/Objective-C/NSTouchBar Catalog/TestViewControllers/PickerViewController.m
/*
See the LICENSE.txt file for this sample’s licensing information.

Abstract:
View controller responsible for showing NSPickerTouchBarItem in an NSTouchBar instance.
*/

#import "PickerViewController.h"

static NSTouchBarCustomizationIdentifier PickerItemCustomizationIdentifier = @"com.TouchBarCatalog.pickerItemController";
static NSTouchBarItemIdentifier PickerItemIdentifier = @"com.TouchBarCatalog.pickerItem";

@interface PickerViewController () <NSTouchBarDelegate>

@property (weak) IBOutlet NSButton *singleSelection;
@property (weak) IBOutlet NSButton *collapsedState;
@property (weak) IBOutlet NSButton *imagesState;

@end

#pragma mark -

@implementation PickerViewController

- (void)viewDidLoad
{
    [super viewDidLoad];
}

- (IBAction)singleSelectionAction:(id)sender {
    // Change each button to the right background color.
    for (NSTouchBarItemIdentifier itemIdentifier in self.touchBar.itemIdentifiers)
    {
        if (itemIdentifier == PickerItemIdentifier)
        {
            NSPickerTouchBarItem *pickerTouchBarItem = [self.touchBar itemForIdentifier:itemIdentifier];
        
            // First reset the selection.
            pickerTouchBarItem.selectedIndex = -1;
            
            pickerTouchBarItem.selectionMode =
                (self.singleSelection.state == NSControlStateValueOn) ?
            NSPickerTouchBarItemSelectionModeSelectOne : NSPickerTouchBarItemSelectionModeSelectAny;
        }
    }
}

- (IBAction)imagesAction:(id)sender {
    // This creates a call to makeTouchBar.
    self.touchBar = nil;
}

- (IBAction)collapsedAction:(id)sender {
    for (NSTouchBarItemIdentifier itemIdentifier in self.touchBar.itemIdentifiers)
    {
        if (itemIdentifier == PickerItemIdentifier)
        {
            NSPickerTouchBarItem *pickerTouchBarItem = [self.touchBar itemForIdentifier:itemIdentifier];
            pickerTouchBarItem.controlRepresentation = self.collapsedState.state ==
                NSControlStateValueOn ? NSPickerTouchBarItemControlRepresentationCollapsed : NSPickerTouchBarItemControlRepresentationAutomatic;
        }
    }
}

- (void)itemPicked:(id)sender {
    NSLog(@"Item Picked = %ld\n", ((NSPickerTouchBarItem *)sender).selectedIndex);
}

#pragma mark NSTouchBarProvider

- (NSTouchBar *)makeTouchBar
{
    NSTouchBar *bar = [[NSTouchBar alloc] init];
    bar.delegate = self;
    
    bar.customizationIdentifier = PickerItemCustomizationIdentifier;
    
    // Set the default ordering of items.
    bar.defaultItemIdentifiers =
        @[PickerItemIdentifier, NSTouchBarItemIdentifierOtherItemsProxy];
    
    bar.customizationAllowedItemIdentifiers = @[PickerItemIdentifier];
    
    return bar;
}

#pragma mark NSTouchBarDelegate

// The system calls this while constructing the NSTouchBar for each NSTouchBarItem you want to create.
- (nullable NSTouchBarItem *)touchBar:(NSTouchBar *)touchBar makeItemForIdentifier:(NSTouchBarItemIdentifier)identifier
{
    if ([identifier isEqualToString:PickerItemIdentifier])
    {
        NSPickerTouchBarItem *pickerTouchBarItem = nil;
        
        NSPickerTouchBarItemSelectionMode selectionMode =
            (self.singleSelection.state == NSControlStateValueOn) ?
                NSPickerTouchBarItemSelectionModeSelectOne : NSPickerTouchBarItemSelectionModeSelectAny;
        // You can also use:
        // selectionMode = NSPickerTouchBarItemSelectionModeSelectMomentary
        
        if (self.imagesState.state == NSControlStateValueOn) {
            NSArray *images = @[ [NSImage imageWithSystemSymbolName:@"applewatch" accessibilityDescription:@"Apple Watch"],
                               [NSImage imageWithSystemSymbolName:@"desktopcomputer" accessibilityDescription:@"iMac"],
                                 [NSImage imageWithSystemSymbolName:@"iphone" accessibilityDescription:@"iPhone"],
                                 [NSImage imageWithSystemSymbolName:@"appletv" accessibilityDescription:@"Apple TV"]
            ];
            pickerTouchBarItem =
                [NSPickerTouchBarItem pickerTouchBarItemWithIdentifier:PickerItemIdentifier
                                                                images:images
                                                         selectionMode:selectionMode
                                                                target:self
                                                                action:@selector(itemPicked:)];
        }
        else
        {
            NSArray *labels = @[@"Item 1", @"Item 2", @"Item 3", @"Item 4"];
            pickerTouchBarItem =
                [NSPickerTouchBarItem pickerTouchBarItemWithIdentifier:PickerItemIdentifier
                                                                labels:labels
                                                         selectionMode:selectionMode
                                                                target:self
                                                                action:@selector(itemPicked:)];
        }
        
        pickerTouchBarItem.selectionColor = [NSColor systemTealColor];
        pickerTouchBarItem.collapsedRepresentationLabel = @"Choices";
        pickerTouchBarItem.controlRepresentation =
            self.collapsedState.state ==
                NSControlStateValueOn ? NSPickerTouchBarItemControlRepresentationCollapsed : NSPickerTouchBarItemControlRepresentationAutomatic;
        
        return pickerTouchBarItem;
    }
    
    return nil;
}

@end

File: appkit-creating-and-customizing-the-touch-bar/Swift/NSTouchBar Catalog/PickerViewController.swift
/*
See the LICENSE.txt file for this sample’s licensing information.

Abstract:
View controller responsible for showing NSPickerTouchBarItem in an NSTouchBar instance.
*/

import Cocoa

class PickerViewController: NSViewController {
    let picker = NSTouchBarItem.Identifier("com.TouchBarCatalog.TouchBarItem.picker")
    let pickerBar = NSTouchBar.CustomizationIdentifier("com.TouchBarCatalog.pickerBar")
    
    @IBOutlet weak var singleSelection: NSButton!
    @IBOutlet weak var collapsedState: NSButton!
    @IBOutlet weak var imagesState: NSButton!
    
    // MARK: - Action Functions
    
    override func viewDidLoad() {
        super.viewDidLoad()
    }
    
    override func makeTouchBar() -> NSTouchBar? {
        let touchBar = NSTouchBar()
        touchBar.delegate = self
        touchBar.customizationIdentifier = pickerBar
        touchBar.defaultItemIdentifiers = [picker]
        return touchBar
    }
    
    @IBAction func singleSelectionAction(_ sender: AnyObject) {
        guard let touchBar = touchBar else { return }
        
        for itemIdentifier in touchBar.itemIdentifiers {
            guard let pickerTouchBarItem = touchBar.item(forIdentifier: itemIdentifier) as? NSPickerTouchBarItem else { continue }
            
            // First reset the selection.
            pickerTouchBarItem.selectedIndex = -1
            
            pickerTouchBarItem.selectionMode = sender.state == NSControl.StateValue.on ? .selectOne : .selectAny
        }
    }
    
    @IBAction func imagesAction(_ sender: AnyObject) {
        // This creates a call to makeTouchBar.
        touchBar = nil
    }
    
    @IBAction func collapsedAction(_ sender: AnyObject) {
        guard let touchBar = touchBar else { return }
        
        for itemIdentifier in touchBar.itemIdentifiers {
            guard let pickerTouchBarItem = touchBar.item(forIdentifier: itemIdentifier) as? NSPickerTouchBarItem else { continue }
            
            pickerTouchBarItem.controlRepresentation =
                sender.state == NSControl.StateValue.on ? .collapsed : .automatic
        }
    }
    
    @objc
    func itemPicked(_ sender: AnyObject) {
        if let picker = sender as? NSPickerTouchBarItem {
            print("\(#function): picker choice \"\(picker.selectedIndex)\"")
        }
    }
}

// MARK: - NSTouchBarDelegate

extension PickerViewController: NSTouchBarDelegate {

    // The system calls this while constructing the NSTouchBar for each NSTouchBarItem you want to create.
    func touchBar(_ touchBar: NSTouchBar, makeItemForIdentifier identifier: NSTouchBarItem.Identifier) -> NSTouchBarItem? {
        switch identifier {
        case picker:
            var pickerTouchBarItem: NSPickerTouchBarItem
            
            var selectionMode: NSPickerTouchBarItem.SelectionMode = .selectAny
            if singleSelection.state == NSControl.StateValue.on {
                selectionMode = .selectOne
            }
            // You can also use:
            // selectionMode = .momentary

            if imagesState.state == NSControl.StateValue.on {
                let images = [ NSImage(systemSymbolName: "applewatch", accessibilityDescription: "Apple Watch")!,
                               NSImage(systemSymbolName: "desktopcomputer", accessibilityDescription: "iMac")!,
                               NSImage(systemSymbolName: "iphone", accessibilityDescription: "iPhone")!,
                               NSImage(systemSymbolName: "appletv", accessibilityDescription: "Apple TV")!]
                pickerTouchBarItem =
                    NSPickerTouchBarItem(identifier: identifier,
                                         images: images,
                                         selectionMode: selectionMode,
                                         target: self,
                                         action: #selector(itemPicked))
            } else {
                let labels = ["Item 1", "Item 2", "Item 3", "Item 4"]
                pickerTouchBarItem =
                    NSPickerTouchBarItem(identifier: identifier,
                                         labels: labels,
                                         selectionMode: selectionMode,
                                         target: self,
                                         action: #selector(itemPicked))
            }

            pickerTouchBarItem.selectionColor = NSColor.systemTeal
            pickerTouchBarItem.collapsedRepresentationLabel = "Choices"
            pickerTouchBarItem.controlRepresentation =
                collapsedState.state == NSControl.StateValue.on ? .collapsed : .automatic

            return pickerTouchBarItem
        default:
            return nil
        }
    }

}
